### ISSUE TITLE
Deadlock during console input and memory read

### ISSUE THREAD

**Issue by f4rnham** [Cisco-Talos/pyrebox#35]
If script callback is reading memory (through volatility or api.r_va) while user is typing to qemu console, deadlock can occur.
Can be consistently reproduced (Win7SP1x86) by running https://github.com/f4rnham/pyrebox/blob/pgd_mismatch/scripts/block_end_test.py

`qemu_global_mutex` and `pyrebox_mutex` mutexes are acquired in opposite order, see traces

```
Thread 1 (Thread 0x7ff3d765aac0 (LWP 16374)):
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007ff3d7244b95 in __GI___pthread_mutex_lock (mutex=0x55cb7fc95e00 <pyrebox_mutex>) at ../nptl/pthread_mutex_lock.c:80   <---------- locked, see T4 #51
#2  0x000055cb7f5fa840 in fd_chr_read (chan=0x55cb84f704e0, cond=<optimized out>, opaque=0x55cb84f702e0) at chardev/char-fd.c:47
#3  0x00007ff3d5a156aa in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#4  0x000055cb7f650a23 in glib_pollfds_poll () at util/main-loop.c:213
#5  os_host_main_loop_wait (timeout=<optimized out>) at util/main-loop.c:261    <---- util/main-loop.c:258 qemu_mutex_lock_iothread(); // locks qemu_global_mutex
#6  main_loop_wait (nonblocking=<optimized out>) at util/main-loop.c:515
#7  0x000055cb7f2c9e5b in main_loop () at vl.c:1919
#8  main (argc=<optimized out>, argv=<optimized out>, envp=<optimized out>) at vl.c:4801
```

```
Thread 4 (Thread 0x7ff3b3a18700 (LWP 16386)):
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007ff3d7244b95 in __GI___pthread_mutex_lock (mutex=mutex@entry=0x55cb7fc6a960 <qemu_global_mutex>) at ../nptl/pthread_mutex_lock.c:80
#2  0x000055cb7f65387a in qemu_mutex_lock (mutex=mutex@entry=0x55cb7fc6a960 <qemu_global_mutex>) at util/qemu-thread-posix.c:65  <---------- locked, see T1 #5
#3  0x000055cb7f306bac in qemu_mutex_lock_iothread () at /root/pyrebox/qemu/cpus.c:1581
#4  0x000055cb7f2ce0e5 in prepare_mmio_access (mr=<optimized out>, mr=<optimized out>) at /root/pyrebox/qemu/exec.c:2885
#5  0x000055cb7f2d31cf in address_space_read_continue (as=as@entry=0x55cb7fc672a0 <address_space_memory>, addr=addr@entry=655360, attrs=..., attrs@entry=...,
    buf=buf@entry=0x7ff3b3a156d0 "x\270\262\320\363\177", len=len@entry=1024, addr1=<optimized out>, l=<optimized out>, mr=0x55cb865585e0) at /root/pyrebox/qemu/exec.c:3009
#6  0x000055cb7f2d342e in address_space_read_full (as=as@entry=0x55cb7fc672a0 <address_space_memory>, addr=655360, addr@entry=94332510433952, attrs=...,
    buf=0x7ff3b3a156d0 "x\270\262\320\363\177", buf@entry=0xa0000 <error: Cannot access memory at address 0xa0000>, len=1024, len@entry=1) at /root/pyrebox/qemu/exec.c:3077
#7  0x000055cb7f2d359e in address_space_read (len=1, buf=0xa0000 <error: Cannot access memory at address 0xa0000>, attrs=..., addr=94332510433952, as=0x55cb7fc672a0 <address_space_memory>)
    at /root/pyrebox/qemu/include/exec/memory.h:1966
#8  address_space_rw (as=as@entry=0x55cb7fc672a0 <address_space_memory>, addr=addr@entry=655360, attrs=..., attrs@entry=..., buf=buf@entry=0x7ff3b3a156d0 "x\270\262\320\363\177",
    len=len@entry=1024, is_write=<optimized out>) at /root/pyrebox/qemu/exec.c:3091
#9  0x000055cb7f2d35d7 in cpu_physical_memory_rw (addr=addr@entry=655360, buf=buf@entry=0x7ff3b3a156d0 "x\270\262\320\363\177", len=len@entry=1024, is_write=is_write@entry=0)
    at /root/pyrebox/qemu/exec.c:3098
#10 0x000055cb7f39277c in cpu_physical_memory_read (len=<optimized out>, buf=0x7ff3b3a156d0, addr=655360) at /root/pyrebox/qemu/include/exec/cpu-common.h:83
#11 connection_read_memory (user_paddr=<optimized out>, buf=buf@entry=0x7ff3644178d0 "\270\006", user_len=user_len@entry=4096) at /root/pyrebox/qemu/pyrebox/qemu_glue.c:915
#12 0x000055cb7f393015 in py_vol_read_memory (dummy=<optimized out>, args=<optimized out>) at /root/pyrebox/qemu/pyrebox/api.cpp:697
#13 0x00007ff3d6958d3c in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    ...
#47 0x00007ff3d6a9a538 in PyEval_EvalCodeEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#48 0x00007ff3d6a6a229 in ?? () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#49 0x00007ff3d6905583 in PyObject_Call () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#50 0x00007ff3d6a99a37 in PyEval_CallObjectWithKeywords () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#51 0x000055cb7f39569b in CallbackManager::deliver_callback (this=<optimized out>, type=type@entry=BLOCK_END_CB, params=...) at /root/pyrebox/qemu/pyrebox/callbacks.cpp:831  <---------- pthread_mutex_lock(&pyrebox_mutex);
#52 0x000055cb7f395a76 in block_end_callback (params=...) at /root/pyrebox/qemu/pyrebox/callbacks.cpp:169
#53 0x000055cb7f396454 in helper_qemu_block_end_callback (cpu=0x55cb85129040, tb=0x7ff3b40293c0 <code_gen_buffer+172950>, from=2189689658, to=0)
    at /root/pyrebox/qemu/pyrebox/qemu_glue_callbacks.c:124
#54 0x00007ff3b40294b2 in code_gen_buffer ()
#55 0x000055cb7f3322ad in cpu_tb_exec (itb=<optimized out>, itb=<optimized out>, cpu=0x7ff3b48bb480 <code_gen_buffer+9159766>) at /root/pyrebox/qemu/accel/tcg/cpu-exec.c:167
#56 cpu_loop_exec_tb (tb_exit=<synthetic pointer>, last_tb=<synthetic pointer>, tb=<optimized out>, cpu=0x7ff3b48bb480 <code_gen_buffer+9159766>)
    at /root/pyrebox/qemu/accel/tcg/cpu-exec.c:579
#57 cpu_exec (cpu=cpu@entry=0x55cb85129040) at /root/pyrebox/qemu/accel/tcg/cpu-exec.c:684
#58 0x000055cb7f3073f2 in tcg_cpu_exec (cpu=<optimized out>) at /root/pyrebox/qemu/cpus.c:1270
#59 qemu_tcg_rr_cpu_thread_fn (arg=<optimized out>) at /root/pyrebox/qemu/cpus.c:1365
#60 0x00007ff3d7242494 in start_thread (arg=0x7ff3b3a18700) at pthread_create.c:333
#61 0x00007ff3d4f75abf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

**Comment by xabiugarte** [comment]
Hi,

I have pushed a commit (b8e7ff2f18e8bda2b00a63e740f929d7fd9f8015) under a different branch: https://github.com/Cisco-Talos/pyrebox/tree/fix_chardev_deadlock

This commit should fix the issue. I have tested it over several use cases, but I will wait a little bit to update the master branch until we make sure it doesn't affect other scripts or other users. The commit adds a long comment explaining the issue, which is left there as documentation.

Let me know if you have any issue with the fix, and once again, thanks for the issues reported and your pull requests.

Regards,
